### PR TITLE
Cap 70 upgrades

### LINKED ISSUE #4739: Add support for SCP Timing Upgrades
Add support for actually upgrading CAP 70 config settings.

### PR DESCRIPTION
**Pull request by SirTyson**
# Description

Resolves #4739

Implements CAP 70. There are a surprising number of places where we need to use the ledger close time. With @marta-lokhova recent PR, there should be a centralized, thread safe network config getter. That being said, we'll still need to run a superclsuter test to see if any of these subsystems are messed up on upgrade boundaries.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/catchup/LedgerApplyManagerImpl.cpp [hunk 1]
```diff
@@ -247,11 +247,12 @@ LedgerApplyManagerImpl::processLedger(LedgerCloseData const& ledgerData,
...
         {
-            auto eta = (catchupTriggerLedger - lastLedgerInBuffer) *
-                       mApp.getConfig().getExpectedLedgerCloseTime();
+            auto etaMS = (catchupTriggerLedger - lastLedgerInBuffer) *
+                         mApp.getLedgerManager().getExpectedLedgerCloseTime(
+                             mApp.getConfig());
             message = fmt::format(
                 FMT_STRING("Waiting for trigger ledger: {:d}/{:d}, ETA: {:d}s"),
-                lastLedgerInBuffer, catchupTriggerLedger, eta.count());
+                lastLedgerInBuffer, catchupTriggerLedger, etaMS.count() / 1000);
```
Comment (bboston7) [inline src/catchup/LedgerApplyManagerImpl.cpp]: Nitpick: Use a `duration_cast` here.

File: src/ledger/LedgerManagerImpl.h [hunk 2]
```diff
@@ -247,10 +251,13 @@ class LedgerManagerImpl : public LedgerManager
...
     uint32_t getLastTxFee() const override;
     uint32_t getLastClosedLedgerNum() const override;
-    SorobanNetworkConfig const& getLastClosedSorobanNetworkConfig() override;
+    SorobanNetworkConfig const&
+    getLastClosedSorobanNetworkConfig() const override;
     SorobanNetworkConfig const& getSorobanNetworkConfigForApply() override;
 
     bool hasLastClosedSorobanNetworkConfig() const override;
+    std::chrono::milliseconds
+    getExpectedLedgerCloseTime(Config const& config) const override;
```
Comment (bboston7) [inline src/ledger/LedgerManagerImpl.h]: Does this need to take a `Config` as an argument? This class has access to the config via `mApp`.

**Review by bboston7 — APPROVED** [review]
Looks great! Just a couple small comments.

File: src/ledger/LedgerManagerImpl.cpp [hunk 3]
```diff
@@ -607,6 +593,35 @@ LedgerManagerImpl::hasLastClosedSorobanNetworkConfig() const
     return mLastClosedLedgerState->hasSorobanConfig();
 }
 
+std::chrono::milliseconds
+LedgerManagerImpl::getExpectedLedgerCloseTime() const
+{
```
Comment (marta-lokhova) [inline src/ledger/LedgerManagerImpl.cpp:635]: releaseAssert(threadIsMain())

File: src/ledger/LedgerManagerImpl.cpp [hunk 4]
```diff
@@ -1373,6 +1388,7 @@ LedgerManagerImpl::setLastClosedLedger(
     LedgerTxn ltx(mApp.getLedgerTxnRoot());
     auto header = ltx.loadHeader();
     header.current() = lastClosed.header;
+    auto lv = header.current().ledgerVersion;
```
Comment (marta-lokhova) [inline src/ledger/LedgerManagerImpl.cpp]: nit: you're accessing an uncommitted half-valid ltx here (but at the right time). It'd be slightly safer to just extract the version from `lastClosed`

File: src/ledger/LedgerManagerImpl.cpp [hunk 5]
```diff
@@ -1381,12 +1397,7 @@ LedgerManagerImpl::setLastClosedLedger(
         advanceBucketListSnapshotAndMakeLedgerState(lastClosed.header, has);
     advanceLastClosedLedgerState(output);
 
-    LedgerTxn ltx2(mApp.getLedgerTxnRoot());
-    auto lv = ltx2.loadHeader().current().ledgerVersion;
-    if (protocolVersionStartsFrom(lv, SOROBAN_PROTOCOL_VERSION))
-    {
-        mApp.getLedgerManager().updateSorobanNetworkConfigForApply(ltx2);
-    }
+    maybeLoadSorobanNetworkConfig(lv);
```
Comment (marta-lokhova) [inline src/ledger/LedgerManagerImpl.cpp:1481]: thanks for this fix!

File: src/ledger/NetworkConfig.cpp [hunk 6]
```diff
@@ -1303,6 +1329,7 @@ SorobanNetworkConfig::loadFromLedger(AbstractLedgerTxn& ltxRoot,
     {
         loadParallelComputeConfig(ltx);
         loadLedgerCostExtConfig(ltx);
+        loadSCPTimingConfig(ltx);
```
Comment (marta-lokhova) [inline src/ledger/NetworkConfig.cpp:1362]: nit: could we use p23 in the if condition for this load? `PARALLEL_SOROBAN_PHASE_PROTOCOL_VERSION` is a bit confusing, since the SCP configs are unrelated

File: src/ledger/NetworkConfig.h [hunk 7]
```diff
@@ -47,6 +47,23 @@ struct MinimumSorobanNetworkConfig
     static constexpr uint32_t STARTING_EVICTION_LEVEL = 1;
 
     static constexpr uint32_t TX_MAX_CONTRACT_EVENTS_SIZE_BYTES = 200;
+
+    // SCP timing minimums
+    static constexpr uint32_t LEDGER_TARGET_CLOSE_TIME_MILLISECONDS = 4000;
```
Comment (marta-lokhova) [inline src/ledger/NetworkConfig.h:52]: What's the rationale behind these ranges? (the CAP says they're approximate and untested). Asking because this limits what the network can do post p23 (and before any other protocol that may change these)

File: src/overlay/SurveyManager.cpp [hunk 8]
```diff
@@ -111,6 +112,10 @@ SurveyManager::startSurveyReporting()
     mCurve25519SecretKey = curve25519RandomSecret();
     mCurve25519PublicKey = curve25519DerivePublic(mCurve25519SecretKey);
 
+    mSurveyThrottleTimeoutMs =
```
Comment (marta-lokhova) [inline src/overlay/SurveyManager.cpp]: No need for this change (so we can keep SURVEY_THROTTLE_TIMEOUT_SEC a const). This throttling const is a very loose heuristic, so I don't think we need the extra complexity.

File: src/scp/test/SCPTests.cpp [hunk 9]
```diff
@@ -308,6 +308,13 @@ class TestSCP : public SCPDriver
     {
         return mSCP.getSlot(slotIndex, false)->getNominationLeaders();
     }
+
+    std::chrono::milliseconds
+    computeTimeout(uint32 roundNumber, bool isNomination) override
```
Comment (marta-lokhova) [inline src/scp/test/SCPTests.cpp:321]: Do we need to test with a few different timeout values for sanity checking? (we've never updated these before, so there might be bugs hiding that rely on the default values)

File: src/simulation/Simulation.cpp [hunk 10]
```diff
@@ -341,6 +341,22 @@ Simulation::stopOverlayTick()
     }
 }
 
+std::chrono::milliseconds
+Simulation::getExpectedLedgerCloseTime() const
```
Comment (marta-lokhova) [inline src/simulation/Simulation.cpp:345]: tbh I'm not certain it's worth grabbing real network configs here at all. This function is typically used to compute timeouts in tests, so just using the default is fine here (maybe just rename to defaultTimeout or something). This is pretty minor though, so up to you.

**Review by marta-lokhova — CHANGES_REQUESTED** [review]
Looks good overall. I added a few minor suggestions. We should make sure we have the right test coverage:
- Now that many parts of the codebase are becoming dependent on the network config, we need to make sure the config is loaded at the right time and up-to-date. For example, some subsystems start before we load or update the state properly, which may cause divergence/crashes. Worth an audit.
- We should run the SCP test suite with a couple of different block times/timeout values. We've never tested SCP with other values, so there might be bugs that never manifested themselves with hardcoded values.
- We should verify that all impacted subsystems behave correctly once new values are introduced. A good task for supercluster (you mentioned doing some simulations in the PR description, any interesting discoveries so far?)

File: src/ledger/NetworkConfig.h (hunk 7 above)
Comment (marta-lokhova) [reply src/ledger/NetworkConfig.h:52]: IOW, should we make these more liberal for flexibility purposes? The actual upgrades will still be conservative and require global coordination of course.

Comment (SirTyson) [reply src/ledger/NetworkConfig.h:52]: This came up in the CAP conversation, but we want these limits to be much narrower than most other config settings. Reason being, if we accidentally upgrade, or a subset of tier 1 does a "rogue upgrade", it's possible to brick the network, such that a follow up "fix" upgrade is not possible if we can't close ledgers. Unlike the other configs, it's pretty easy to reverse them, but there's not much you can do if you can't close blocks. These are all untested at this point, so we made the bounds very conservative to start. The idea is once we have more testing and perf work, if p24 is capable of lower blocktimes than 4 seconds, we lower the limit with the p24 release. 

For testing purposes, perhaps we can set different bounds via ifdef for ssc testing.

File: src/simulation/Simulation.cpp (hunk 10 above)
Comment (SirTyson) [reply src/simulation/Simulation.cpp:345]: I think this is best. Reason being, in the future if we move the default close time down from 5 seconds, this gives you better coverage. There were also a few tests that were using the wrong close time, such as the acceptance tests in `CoreTests.cpp`, and this function cleans that up.

Comment (marta-lokhova) [reply src/simulation/Simulation.cpp:345]: what I was suggesting is that we simply use 5 seconds, since we only use this function for timeout computation, unsure how that would impact coverage. We only use these for timeouts, don't we? Also, which acceptance test was wrong?

Comment (SirTyson) [reply src/simulation/Simulation.cpp:345]: I think it could actually have some test impact. For example, if we're testing something like "it should happen within 5 ledgers" in simulation, we'll often crank for 5 * ledger_close_time. If your ledger close time is wrong, you're not actually testing what you think you are.

"3 nodes 2 running threshold 2" and I think all the other simulation tests in "CoreTests.cpp" were wrong in this way. Our timeouts were using the Herder defined 5 second const value, but had actually set `ARTIFICIALLY_ACCELERATE_TIME_FOR_TESTING` in the simulated nodes. So all the lines that were `nLedgers * Herder::EXP_LEDGER_TIMESPAN_SECONDS` were semantically wrong, since the simulation ledger close time wasn't actually `Herder::EXP_LEDGER_TIMESPAN_SECONDS`.

I think this is a pretty simple change anyway and cleans up our tests, as it's not super clear if `ARTIFICIALLY_ACCELERATE_TIME_FOR_TESTING` is set or not since Simulation sets it automatically if it doesn't get a config in it's constructor.

File: src/overlay/SurveyManager.cpp (hunk 8 above)
Comment (SirTyson) [reply src/overlay/SurveyManager.cpp]: Hmm it's actually been pretty challenging to get tests to work with the 5 second value. The tests all run simulation at 1 second close time, and I think the changes to tests are more complex than the actual survey change. Worst case we can address this as a follow up issue but I think it's kinda non-trivial since `mApp.getConfig().getExpectedLedgerCloseTime()` previously returned 1 second and not 5 for these tests.

File: src/simulation/Simulation.cpp (hunk 10 above)
Comment (marta-lokhova) [reply src/simulation/Simulation.cpp:345]: Relying on this timeout in tests for correctness purposes sounds wrong: the actual success condition depends on the lambda passed to `crankUntil`. The timeout is there to prevent the test from cranking forever in case the condition isn't met, and typically it's way longer than the expected completion time. I don't mind keeping it, but I want to be clear that using a conservative 5 second default does not impact test correctness (e.g. in CoreTests you'd just crank for a bit longer in case of a failure).

File: src/scp/test/SCPTests.cpp (hunk 9 above)
Comment (SirTyson) [reply src/scp/test/SCPTests.cpp:321]: Added a matrix of different timeouts values to the SCP unit tests.

**Comment by SirTyson** [comment]
I've added unit tests to SCP for different timeout values. While auditing the startup path in particular, it seems like `Maintainer` can definitely call this function prior to last closed ledger state being set. I've but a check in that if `mLastClosedLedgerState` is not set, we just return the default value instead of asserting. I think this is the safest action. Any sort of concensus related functionality that has a strong dependency on trigger time will call the function after `mLastClosedLedgerState` is populated anyway. Some services want a general sense of ledger timing, but don't really need the exact value, such as `Maintainer`.

File: src/ledger/NetworkConfig.h (hunk 7 above)
Comment (SirTyson) [reply src/ledger/NetworkConfig.h:52]: I actually ended up changing these bounds slightly due to some unit tests values at the lower bounds. I think these are pretty safe for the release, but we'll need ssc testing for better understanding for actual upgrades and future bounds.

File: src/overlay/SurveyManager.cpp (hunk 8 above)
Comment (marta-lokhova) [reply src/overlay/SurveyManager.cpp]: Ok. This override is in the wrong place though (throttling begins before the reporting phase). So it's pretty confusing that the value changes mid-survey. I think the root cause is that `TARGET_LEDGER_CLOSE_TIME_BEFORE_PROTOCOL_VERSION_23_MS` doesn't handle accelerated time mode. You can solve that by adding a helper method to Config that returns either 5 or 1 second. Then the override can be removed.

File: src/ledger/NetworkConfig.h (hunk 7 above)
Comment (marta-lokhova) [reply src/ledger/NetworkConfig.h:52]: Makes sense. It's fine to start conservative, but can you elaborate on the failure mode here (network unable to close)?

> I actually ended up changing these bounds slightly due to some unit tests values at the lower bounds. I think these are pretty safe for the release, but we'll need ssc testing for better understanding for actual upgrades and future bounds.

Can you elaborate on the specific changes? It sounds a bit suspicious to change prod values based on tests, unless the test pointed out that the initial values were wrong?

File: src/ledger/LedgerManagerImpl.cpp [hunk 11]
```diff
@@ -618,6 +604,48 @@ LedgerManagerImpl::hasLastClosedSorobanNetworkConfig() const
...
+    }
+#endif
+
+    // If a service calls this function during startup, it's possible that we
+    // have not yet set last closed ledger state, so return the "default" value.
+    // This is fine, since any protocol or consensus related functions that have
+    // a strong requirement that the value returned is the actual version in
+    // the network config will always be called after we've set the last closed
+    // ledger state.
+    if (!mLastClosedLedgerState)
```
Comment (marta-lokhova) [inline src/ledger/LedgerManagerImpl.cpp]: Could we avoid doing this? This is pretty error-prone, as the callers will have no idea they might be dealing with an incorrect value. Instead, this function should enforce that mLastClosedLedgerState is set, and let the callers deal with scenarios where for whatever reason it's not (or crash to highlight a bug). In case of Maintainer, we start it after we load the LCL, so I think it should be fine.

File: src/scp/test/SCPTests.cpp (hunk 9 above)
Comment (marta-lokhova) [reply src/scp/test/SCPTests.cpp:321]: thanks! The matrix might be too aggressive though, as it causes a combinatorial explosion of SECTIONs (it shows up in our CI too, the build for this PR takes ~1.5 hours). We can reduce this to 3 combinations: current network config, minimum network config and maximum network config.

File: src/ledger/NetworkConfig.h (hunk 7 above)
Comment (SirTyson) [reply src/ledger/NetworkConfig.h:52]: The minimum timeout increments went from 500 ms to 750 ms. There's probably not something fundamentally different between minimum values of 500 ms to 750 ms, but unit tests failed at 500. Given that these are pretty arbitrary values that haven't been tested in SSC, and given that we don't actually plan on changing timeout values any time soon (these are included for completeness, target ledger close time is what we're actually planning on upgrading), I think having more conservative limits is fine. We'll follow up with SSC tests where we can actually simulate values better. I would put basically 0 stock in the "prod" values since these are basically just made up at this point, but again, we don't actually plan on upgrading, we just needed some bounds.

File: src/overlay/SurveyManager.cpp (hunk 8 above)
Comment (SirTyson) [reply src/overlay/SurveyManager.cpp]: I've added a function that looks if test config flags are set. I decided on a local function just for survey manager instead of putting something in config. I don't want there to be 2 places globally to get the ledger close time, as this could be a potential footgun later on.

**Comment by SirTyson** [comment]
Rebased and addressed comments

File: src/ledger/LedgerManagerImpl.cpp [hunk 12]
```diff
@@ -872,6 +858,37 @@ LedgerManagerImpl::hasLastClosedSorobanNetworkConfig() const
     return mLastClosedLedgerState->hasSorobanConfig();
 }
 
+std::chrono::milliseconds
+LedgerManagerImpl::getExpectedLedgerCloseTime() const
+{
+    releaseAssert(threadIsMain());
+
+#ifdef BUILD_TESTS
+    auto const& cfg = mApp.getConfig();
```
Comment (dmkozh) [inline src/ledger/LedgerManagerImpl.cpp:639]: This block is duplicated in `SurveyManager`. I think this deserves a helper function.

**Review by marta-lokhova — APPROVED** [review]

**Comment by marta-lokhova** [comment]
hmm, CI is still quite a bit slower (~1.5hrs), but we can address this as part of https://github.com/stellar/stellar-core-internal/issues/371 given the timeline

File: src/ledger/LedgerManagerImpl.cpp (hunk 12 above)
Comment (SirTyson) [reply src/ledger/LedgerManagerImpl.cpp:639]: Added a helper in `Config`

**Review by marta-lokhova — DISMISSED** [review]

**Review by marta-lokhova — DISMISSED** [review]

**Review by dmkozh — APPROVED** [review]

**Review by dmkozh — APPROVED** [review]
